### PR TITLE
remove unneeded rw_lock_t debug info

### PR DESCRIPTION
**Pull request by kevgs**
```c++
WARNING: ThreadSanitizer: data race (pid=6798)
Write of size 4 at 0x7b30000007e0 by thread T22:
#0 rw_lock_s_lock_low(rw_lock_t*, unsigned long, char const*, unsigned int) storage/innobase/include/sync0rw.ic:253:20 (mysqld+0x210986d)
#1 rw_lock_s_lock_func(rw_lock_t*, unsigned long, char const*, unsigned int) storage/innobase/include/sync0rw.ic:288:7 (mysqld+0x2109724)
#2 pfs_rw_lock_s_lock_func(rw_lock_t*, unsigned long, char const*, unsigned int) storage/innobase/include/sync0rw.ic:648:3 (mysqld+0x210916a)
#3 row_purge_parse_undo_rec(purge_node_t*, unsigned char*, bool*, que_thr_t*) storage/innobase/row/row0purge.cc:928:2 (mysqld+0x2107d34)
#4 row_purge(purge_node_t*, unsigned char*, que_thr_t*) storage/innobase/row/row0purge.cc:1102:10 (mysqld+0x2106170)
#5 row_purge_step(que_thr_t*) storage/innobase/row/row0purge.cc:1184:3 (mysqld+0x2105e4c)
#6 que_thr_step(que_thr_t*) storage/innobase/que/que0que.cc:1054:9 (mysqld+0x2002735)
#7 que_run_threads_low(que_thr_t*) storage/innobase/que/que0que.cc:1116:14 (mysqld+0x2000b00)
#8 que_run_threads(que_thr_t*) storage/innobase/que/que0que.cc:1156:2 (mysqld+0x20006ca)
#9 srv_task_execute() storage/innobase/srv/srv0srv.cc:2580:3 (mysqld+0x21d5d84)
#10 srv_worker_thread storage/innobase/srv/srv0srv.cc:2627:7 (mysqld+0x21d5947)

Previous write of size 4 at 0x7b30000007e0 by thread T21:
#0 rw_lock_s_lock_low(rw_lock_t*, unsigned long, char const*, unsigned int) storage/innobase/include/sync0rw.ic:253:20 (mysqld+0x210986d)
#1 rw_lock_s_lock_func(rw_lock_t*, unsigned long, char const*, unsigned int) storage/innobase/include/sync0rw.ic:288:7 (mysqld+0x2109724)
#2 pfs_rw_lock_s_lock_func(rw_lock_t*, unsigned long, char const*, unsigned int) storage/innobase/include/sync0rw.ic:648:3 (mysqld+0x210916a)
#3 row_purge_parse_undo_rec(purge_node_t*, unsigned char*, bool*, que_thr_t*) storage/innobase/row/row0purge.cc:928:2 (mysqld+0x2107d34)
#4 row_purge(purge_node_t*, unsigned char*, que_thr_t*) storage/innobase/row/row0purge.cc:1102:10 (mysqld+0x2106170)
#5 row_purge_step(que_thr_t*) storage/innobase/row/row0purge.cc:1184:3 (mysqld+0x2105e4c)
#6 que_thr_step(que_thr_t*) storage/innobase/que/que0que.cc:1054:9 (mysqld+0x2002735)
#7 que_run_threads_low(que_thr_t*) storage/innobase/que/que0que.cc:1116:14 (mysqld+0x2000b00)
#8 que_run_threads(que_thr_t*) storage/innobase/que/que0que.cc:1156:2 (mysqld+0x20006ca)
#9 srv_task_execute() storage/innobase/srv/srv0srv.cc:2580:3 (mysqld+0x21d5d84)
#10 srv_worker_thread storage/innobase/srv/srv0srv.cc:2627:7 (mysqld+0x21d5947)


WARNING: ThreadSanitizer: data race (pid=6798)
Write of size 8 at 0x7b30000007d0 by thread T22:
#0 rw_lock_s_lock_low(rw_lock_t*, unsigned long, char const*, unsigned int) storage/innobase/include/sync0rw.ic:252:25 (mysqld+0x210982c)
#1 rw_lock_s_lock_func(rw_lock_t*, unsigned long, char const*, unsigned int) storage/innobase/include/sync0rw.ic:288:7 (mysqld+0x2109724)
#2 pfs_rw_lock_s_lock_func(rw_lock_t*, unsigned long, char const*, unsigned int) storage/innobase/include/sync0rw.ic:648:3 (mysqld+0x210916a)
#3 row_purge_parse_undo_rec(purge_node_t*, unsigned char*, bool*, que_thr_t*) storage/innobase/row/row0purge.cc:928:2 (mysqld+0x2107d34)
#4 row_purge(purge_node_t*, unsigned char*, que_thr_t*) storage/innobase/row/row0purge.cc:1102:10 (mysqld+0x2106170)
#5 row_purge_step(que_thr_t*) storage/innobase/row/row0purge.cc:1184:3 (mysqld+0x2105e4c)
#6 que_thr_step(que_thr_t*) storage/innobase/que/que0que.cc:1054:9 (mysqld+0x2002735)
#7 que_run_threads_low(que_thr_t*) storage/innobase/que/que0que.cc:1116:14 (mysqld+0x2000b00)
#8 que_run_threads(que_thr_t*) storage/innobase/que/que0que.cc:1156:2 (mysqld+0x20006ca)
#9 srv_task_execute() storage/innobase/srv/srv0srv.cc:2580:3 (mysqld+0x21d5d84)
#10 srv_worker_thread storage/innobase/srv/srv0srv.cc:2627:7 (mysqld+0x21d5947)

Previous write of size 8 at 0x7b30000007d0 by thread T20:
#0 rw_lock_s_lock_low(rw_lock_t*, unsigned long, char const*, unsigned int) storage/innobase/include/sync0rw.ic:252:25 (mysqld+0x210982c)
#1 rw_lock_s_lock_func(rw_lock_t*, unsigned long, char const*, unsigned int) storage/innobase/include/sync0rw.ic:288:7 (mysqld+0x2109724)
#2 pfs_rw_lock_s_lock_func(rw_lock_t*, unsigned long, char const*, unsigned int) storage/innobase/include/sync0rw.ic:648:3 (mysqld+0x210916a)
#3 row_purge_parse_undo_rec(purge_node_t*, unsigned char*, bool*, que_thr_t*) storage/innobase/row/row0purge.cc:928:2 (mysqld+0x2107d34)
#4 row_purge(purge_node_t*, unsigned char*, que_thr_t*) storage/innobase/row/row0purge.cc:1102:10 (mysqld+0x2106170)
#5 row_purge_step(que_thr_t*) storage/innobase/row/row0purge.cc:1184:3 (mysqld+0x2105e4c)
#6 que_thr_step(que_thr_t*) storage/innobase/que/que0que.cc:1054:9 (mysqld+0x2002735)
#7 que_run_threads_low(que_thr_t*) storage/innobase/que/que0que.cc:1116:14 (mysqld+0x2000b00)
#8 que_run_threads(que_thr_t*) storage/innobase/que/que0que.cc:1156:2 (mysqld+0x20006ca)
#9 srv_task_execute() storage/innobase/srv/srv0srv.cc:2580:3 (mysqld+0x21d5d84)
#10 srv_worker_thread storage/innobase/srv/srv0srv.cc:2627:7 (mysqld+0x21d5947)
```

I'm contributing this new code of the whole pull request, including one or several files that are either new files or modified ones, under the BSD-new license.

### CODE REVIEW

File: storage/innobase/include/sync0rw.h [hunk 1]
```diff
@@ -615,12 +615,12 @@ struct rw_lock_t
...
 	unsigned	is_block_lock:1;
 
-	/** Line number where last time s-locked */
-	unsigned	last_s_line:14;
-
 	/** Line number where last time x-locked */
 	unsigned	last_x_line:14;
 
+	/** Line number where last time s-locked */
+	ulint		last_s_line;
```
Comment (svoj) [inline storage/innobase/include/sync0rw.h]: No uliint please, use uint32_t instead.

File: storage/innobase/sync/sync0arr.cc [hunk 2]
```diff
@@ -590,7 +590,7 @@ sync_array_cell_print(
 				"number of readers " ULINTPF
 				", waiters flag %u, "
 				"lock_word: " ULINTPFx "\n"
-				"Last time read locked in file %s line %u\n"
+				"Last time read locked in file %s line " ULINTPF "\n"
```
Comment (svoj) [inline storage/innobase/sync/sync0arr.cc]: Please use atomic load both for last_s_line and last_s_file_name. There should be 4 occurrences total.

**Comment by kevgs** [comment]
4 atomic loads total.

**Comment by svoj** [comment]
Looks good, thanks! I'll get this merged soon.

**Comment by svoj** [comment]
We'll most probably need to add casts for Windows. I'll need to get this through buildbot first.

**Comment by kevgs** [comment]
Does Windows require explicit casts between signed and unsigned integers of the same type?

**Comment by svoj** [comment]
That's because we have rather silly atomic stuff on Windows:
```
static inline int32 my_atomic_load32(int32 volatile *a)
{
  return (int32)InterlockedCompareExchange((volatile LONG *)a, 0, 0);
}
```

So it is declared to get signed and return signed 32-bit int.

**Comment by svoj** [comment]
@dr-m could you take care of this PR?

File: storage/innobase/sync/sync0arr.cc [hunk 3]
```diff
@@ -1403,8 +1396,6 @@ sync_arr_fill_sys_semphore_waits_table(
 						OK(field_store_ulint(fields[SYS_SEMAPHORE_WAITS_READERS], rw_lock_get_reader_count(rwlock)));
 						OK(field_store_ulint(fields[SYS_SEMAPHORE_WAITS_WAITERS_FLAG], (longlong)rwlock->waiters));
 						OK(field_store_ulint(fields[SYS_SEMAPHORE_WAITS_LOCK_WORD], (longlong)rwlock->lock_word));
-						OK(field_store_string(fields[SYS_SEMAPHORE_WAITS_LAST_READER_FILE], innobase_basename(my_atomic_loadptr_explicit(&rwlock->last_s_file_name, MY_MEMORY_ORDER_RELAXED))));
-						OK(fields[SYS_SEMAPHORE_WAITS_LAST_READER_LINE]->store(my_atomic_load32_explicit(&rwlock->last_s_line, MY_MEMORY_ORDER_RELAXED), true));
 						fields[SYS_SEMAPHORE_WAITS_LAST_READER_LINE]->set_notnull();
```
Comment (dr-m) [inline storage/innobase/sync/sync0arr.cc]: I think that both `SYS_SEMAPHORE_WAITS_LAST_READER_LINE` and `SYS_SEMAPHORE_WAITS_LAST_READER_FILE` and related `INFORMATION_SCHEMA.INFORMATION_SCHEMA.INNODB_SYS_SEMAPHORE_WAITS` table columns must be removed. This removal must be mentioned in the commit message too.

**Comment by kevgs** [comment]
Commit message:
```
remove unneeded rw_lock_t debug info

rw_lock_t: remove last_s_file_name, last_s_line members

information_schema.sys_semaphore_waits: remove LAST_READER_LINE, LAST_READER_FILE
```

**Comment by svoj** [comment]
Thanks!

**Pull request merged**